File: backend/models/message.py
"""
Message data model.

This module defines the Message model for chat operations.
"""

from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field
import uuid

from .emotions import Emotions


class Message(BaseModel):
    """Message model representing a chat message."""

    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    role: Literal["user", "assistant"]
    content: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    character_id: Optional[str] = Field(default=None, description="ID of the character this message belongs to")
    character_name: Optional[str] = Field(default=None, description="Name of the character (for group chat display)")
    emotions: Optional[Emotions] = Field(default=None, description="Detected emotions (assistant messages only)")


class MessageCreate(BaseModel):
    """Model for creating a new message."""

    content: str


class MessageResponse(BaseModel):
    """Model for message response containing both user and assistant messages."""

    user_message: Message
    assistant_message: Message


class MessagesResponse(BaseModel):
    """Model for paginated messages response."""

    messages: list[Message]
    total: int

File: frontend/src/types/message.ts
/**
 * Type definitions for Message-related data structures.
 */

/**
 * Emotions interface representing character's emotional state.
 * All values are in range 0-100.
 */
export interface Emotions {
  /** Fear level (0-100) */
  fear: number;
  /** Anger level (0-100) */
  anger: number;
  /** Sadness level (0-100) */
  sadness: number;
  /** Disgust level (0-100) */
  disgust: number;
  /** Joy level (0-100) */
  joy: number;
}

export interface Message {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  created_at: string;
  character_id?: string;
  /** Optional character name for group chat display */
  character_name?: string;
  /** Optional emotions data, only present in assistant messages */
  emotions?: Emotions;
}

export interface MessageCreate {
  content: string;
}

export interface MessageResponse {
  user_message: Message;
  assistant_message: Message;
}

export interface MessagesResponse {
  messages: Message[];
  total: number;
}
